x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/api_key/40_role_descriptors.yml

```diff
@@ -0,0 +1,332 @@
+                    "privileges": ["all"]
+                  }
+                ]
+              }
+            }
+          }
+  - match: { name: "key-3-with-multiple-assigned-role-descriptors" }
+  - is_true: id
+  - set: { id: api_key_id_3 }
+
+  - do:
+      headers:
+        Authorization: "Basic YXBpX2tleV91c2VyOngtcGFjay10ZXN0LXBhc3N3b3Jk" # api_key_user
+      security.get_api_key:
+        id: "$api_key_id_3"
+  - match: { "api_keys.0.id": "$api_key_id_3" }
+  - match: { "api_keys.0.role_descriptors": {
+    "role-a": {
+      "cluster": [
+        "all"
+      ],
+      "indices": [
+        {
+          "names": [
+            "index-a"
+          ],
+          "privileges": [
+            "read"
+          ],
+          "allow_restricted_indices": false
+        }
+      ],
+      "applications": [ ],
+      "run_as": [ ],
+      "metadata": { },
+      "transient_metadata": {
+        "enabled": true
+      }
+    },
+    "role-b": {
+      "cluster": [
+        "manage"
+      ],
+      "indices": [
+        {
+          "names": [
+            "index-b"
+          ],
+          "privileges": [
+            "all"
+          ],
+          "allow_restricted_indices": false
+        }
+      ],
+      "applications": [ ],
+      "run_as": [ ],
+      "metadata": { },
+      "transient_metadata": {
+        "enabled": true
+      }
+    }
+  }
+  }
+
+  # Query API keys
+  - do:
+      headers:
+        Authorization: "Basic YXBpX2tleV91c2VyOngtcGFjay10ZXN0LXBhc3N3b3Jk" # api_key_user
+      security.query_api_keys:
+        body: >
+          {
+            "sort": [ "name" ]
+          }
+  - match: { total: 4 }
+  - match: { count: 4 }
+  - match: { "api_keys.0.name": "key-0-with-implicit-inherit" }
+  - match: { "api_keys.0.role_descriptors": { } }
+  - match: { "api_keys.1.name": "key-1-with-explicit-inherit" }
+  - match: { "api_keys.1.role_descriptors": { } }
+  - match: { "api_keys.2.name": "key-2-with-single-assigned-role-descriptor"}
+  - match: { "api_keys.2.role_descriptors": {
+    "role-a": {
+      "cluster": [
+        "all"
+      ],
+      "indices": [
+        {
+          "names": [
+            "index-a"
+          ],
+          "privileges": [
+            "read"
+          ],
+          "allow_restricted_indices": false
+        }
+      ],
+      "applications": [ ],
+      "run_as": [ ],
+      "metadata": { },
+      "transient_metadata": {
+        "enabled": true
+      }
+    }
+  }
+  }
+  - match: { "api_keys.3.name": "key-3-with-multiple-assigned-role-descriptors"}
+  - match: { "api_keys.3.role_descriptors": {
+    "role-a": {
+      "cluster": [
+        "all"
+      ],
+      "indices": [
+        {
+          "names": [
+            "index-a"
+          ],
+          "privileges": [
+            "read"
+          ],
+          "allow_restricted_indices": false
+        }
+      ],
+      "applications": [ ],
+      "run_as": [ ],
+      "metadata": { },
+      "transient_metadata": {
```

Comment:
`transient_metadata` feels more like an internal value as opposed to something that would be part of the public API, but we already expose it via the roles API as well so it makes sense to stay consistent.

Reply:
No it is intentional for `transient_metadata` to be in the response, at least for regular roles. When a role has DLS/FLS but the cluster license does not allow it, the `transient_metadata` will have this information to let user know that the role is disabled and for what reason, e.g.:
```
"transient_metadata": {
  "unlicensed_features": ["dls", "fls"],
  "enabled": false
}
```
That said, the license enforcement for API key's permission works differently and does not rely on the transient metadata. But for future proof and consistent, I'd say it's better to keep it in the response.

Reply:
Got it. I didn't realize this was meant for end-users. Makes sense to keep it in the response 👍